File: PI2CabelloLinaresAntonio(C)/ejercicios/ejercicio1Iterativo.c
/*
 * ejercicio1Iterativo.c
 *
 *  Created on: 19 nov. 2020
 *      Author: anton
 */

#include "ejercicio1Iterativo.h"

int cadenasIterativo(char *cadena1, char *cadena2) {
	//las dos cadenas son del mismo tamaño
	int izquierda = 0;
	int derecha = strlen(cadena1);
	int posicion = 0; //entero que me indicara la posicion
	int it = (izquierda + derecha) / 2; //elemento del medio de la cadena

	while (izquierda <= derecha) {

		//establezco dos tipos de filtros, uno para cuando el elemento del medio
		//es igual en las dos cadenas y otro para cuando es distinto

		if (cadena1[it] == cadena2[it]) {

			if (cadena1[it] != cadena2[it]) {	//si iterando las cadenas por el principio y hacia la izquierda encuentro un caracter distintos(comparando las dos cadenas)
				posicion = it + 1;//obtengo la posicion mas 1 pq empieza a iterar desde 0
				return posicion;
			}

			else {		//sino, vuelvo a repetir la iteracion
				izquierda = it;//se va dividiendo el problema hasta encontrar la solucion
			}

		}

		else if (cadena1[it] != cadena2[it]) {

			if (cadena1[it - 1] == cadena2[it - 1]) {	//si iterando las cadenas por el principio y hacia la izquierda encuentro un caracter distintos(comparando las dos cadenas)
				posicion = it;
				return posicion;
			}

			else {
				derecha = it;
			}

		}

		it = (izquierda + derecha) / 2;
	}

	return -1;

}

File: PI2CabelloLinaresAntonio/src/ejercicios/ejercicio1Iterativo.java
package ejercicios;



public class ejercicio1Iterativo {

	public static Integer cadenasIterativo(String cadena1, String cadena2) {
		Integer izquierda = 0;
		Integer derecha = cadena1.length();

		Integer posicion = 0;

		Integer it = (izquierda + derecha) / 2;

		while (izquierda <= derecha) {

			// establezco dos tipos de filtros, uno para cuando el elemento del medio
			// es igual en las dos cadenas y otro para cuando es distinto

			if (cadena1.charAt(it) == cadena2.charAt(it)) {
				// Si el elemento del medio es igual voy a buscar el distinto empezando por el
				// medio y yendo para la derecha
				if (cadena1.charAt(it) != cadena2.charAt(it)) { // si iterando las cadenas por el
																				// principio
																				// encuentro un caracter
																				// distinto(comparando las dos cadenas)

					posicion = it + 1; // obtengo la posicion del elemento por la derecha
					
					return posicion;
				}

				else {// sino vuelvo a repetir la iteracion
					izquierda = it;// esta linea repite el problema dividiendolo en 2 cada vez mas
				}
			}

			else if (cadena1.charAt(it) != cadena2.charAt(it)) {
				//si el elemento que me encuentro en medio es distinto en las dos cadenas, estará en la izquierda (it-1)

				if (cadena1.charAt(it - 1) == cadena2.charAt(it - 1)) {// si iterando las cadenas por el
																						// principio y hacia la
																						// izquierda encuentro un
																						// caracter distintos(comparando
																						// las dos cadenas)
					posicion = it;
					return posicion;
				}

				else {
					derecha = it;
				}
			}

			it = (izquierda + derecha) / 2;

		}

		return -1;

	}


}
